File: src/identity.ts
import os from "node:os";
import path from "node:path";
import { unlink } from "node:fs/promises";
import { DEFAULT_STATE_DIR, readJsonFile, writeJsonAtomic } from "./state.js";

export type LocalDeviceCredential = {
  credentialId: string;
  credentialSecret: string;
  deviceId: string;
  deviceName: string;
  serverUrl: string;
  createdAt: number;
};

const CREDENTIAL_FILE = process.env.CODELOCAL_CREDENTIAL_FILE ?? path.join(DEFAULT_STATE_DIR, "device-credential.json");

export async function loadLocalCredential(serverUrl?: string) {
  const value = await readJsonFile<LocalDeviceCredential | null>(CREDENTIAL_FILE, null);
  if (!value) return null;
  if (serverUrl && value.serverUrl.replace(/\/$/, "") !== serverUrl.replace(/\/$/, "")) return null;
  return value;
}

export async function saveLocalCredential(value: Omit<LocalDeviceCredential, "createdAt"> & { createdAt?: number }) {
  const credential: LocalDeviceCredential = {
    ...value,
    createdAt: value.createdAt ?? Date.now(),
  };
  await writeJsonAtomic(CREDENTIAL_FILE, credential);
  return credential;
}

export async function deleteLocalCredential() {
  await unlink(CREDENTIAL_FILE).catch(() => undefined);
}

export function defaultDeviceIdentity() {
  return {
    deviceId: process.env.CODELOCAL_DEVICE_ID ?? os.hostname(),
    deviceName: process.env.CODELOCAL_DEVICE_NAME ?? os.hostname(),
  };
}

File: internal/identity/identity.go
package identity

import (
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/0xmarkhydra/codelocal/internal/state"
)

type Credential struct {
	CredentialID     string `json:"credentialId"`
	CredentialSecret string `json:"credentialSecret"`
	DeviceID         string `json:"deviceId"`
	DeviceName       string `json:"deviceName"`
	ServerURL        string `json:"serverUrl"`
	CreatedAt        int64  `json:"createdAt"`
}

func credentialFile() string {
	if value := os.Getenv("CODELOCAL_CREDENTIAL_FILE"); value != "" {
		return value
	}
	return filepath.Join(state.Dir(), "device-credential.json")
}

func Load(serverURL string) (*Credential, error) {
	var out Credential
	if err := state.ReadJSON(credentialFile(), &out); err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	if serverURL != "" && strings.TrimRight(out.ServerURL, "/") != strings.TrimRight(serverURL, "/") {
		return nil, nil
	}
	return &out, nil
}

func Save(value Credential) error {
	if value.CreatedAt == 0 {
		value.CreatedAt = time.Now().UnixMilli()
	}
	return state.WriteJSONAtomic(credentialFile(), value)
}

func Delete() error {
	err := os.Remove(credentialFile())
	if os.IsNotExist(err) {
		return nil
	}
	return err
}

func Device() (id, name string) {
	host, _ := os.Hostname()
	id = os.Getenv("CODELOCAL_DEVICE_ID")
	if id == "" {
		id = host
	}
	name = os.Getenv("CODELOCAL_DEVICE_NAME")
	if name == "" {
		name = host
	}
	return id, name
}
